src/Functions/zCurveBase.h

```diff
@@ -0,0 +1,349 @@
+            size_t initial_shift = number_of_bits - (arity << 1) + argument_index;
+            ResultType get_block = max_block_value << initial_shift;
+            ResultType plus_one = 1ull << initial_shift;
+
+            for (size_t i = 0; i <= initial_shift / arity; ++i)
+            {
+                // Handling the next block for MAX
+                left_block_max = left_max & get_block;
+                right_block_max = right_max & get_block;
+                is_left_best = static_cast<bool>(left_block_max & plus_one);
+                if (left_block_max < right_block_max)
+                {
+                    if (!is_left_best)
+                    {
+                        left_max += plus_one;
+                        left_max &= ~(plus_one - 1);
+                    }
+                    if (is_left_best || left_block_max + plus_one < right_block_max)
+                    {
+                        right_max |= plus_one - 1;
+                    }
+                }
+                else
+                {
+                    left_max |= get_block;
+                    left_max ^= get_block;
+                    left_max |= right_block_max;
+                }
+
+                // Handling the next block for MIN
+                left_block_min = left_min & get_block;
+                right_block_min = right_min & get_block;
+                is_right_worse = static_cast<bool>(right_block_min & plus_one);
+                if (left_block_min < right_block_min)
+                {
+                    if (is_right_worse)
+                    {
+                        right_min -= plus_one;
+                        right_min |= (plus_one - 1);
+                    }
+                    if (!is_right_worse || left_block_min + plus_one < right_block_min)
+                    {
+                        left_min &= ~(plus_one - 1);
+                    }
+                }
+                else
+                {
+                    right_min |= get_block;
+                    right_min ^= get_block;
+                    right_min |= left_block_min;
+                }
+                get_block >>= arity;
+                plus_one >>= arity;
+            }
+        }
+
+        return {extractArgument(right_min, argument_index, arity),
+                extractArgument(left_max, argument_index, arity)};
+    }
+    /// Put the bits of argument at every arity-th position in the resulting number
+    ResultType zShiftElement(
+            const char * argument,
+            size_t argument_size,
+            size_t arity,
+            const DataTypePtr & type) const
+    {
+        static constexpr int byte_length = sizeof(ResultType), bit_length = byte_length << 3;
+        ResultType tmp = 0;
+        memcpy(&tmp, argument, argument_size);
+        Op::encode(tmp, type);
+        ResultType result = 0;
+        for (int bit = bit_length - 1, curr = bit_length - 1; curr >= 0; --bit, curr -= arity)
```

Comment:
@stavrolia @alexey-milovidov @TCeason @bakey 
Can not understand the implementation here.
Per my understanding, the z-order value should be interleaved by the bits of different arguments in order.
But here, it seems that the result will be only set as 1 at every arity-th position.

Reply:
Indeed, this function takes a single argument and basically interleaves its bits with (arity - 1) zeroes. Later we take these values for each argument, shift them accordingly to the argument's index (so that its bits end up at the correct positions) and xor the resulting numbers together. 
I don't remember why this was written this way, I think I thought this was a good way to separate different stages since there are a lot of things going on at the same time.

Reply:
yes,I agree with  with that

Reply:
> Indeed, this function takes a single argument and basically interleaves its bits with (arity - 1) zeroes. Later we take these values for each argument, shift them accordingly to the argument's index (so that its bits end up at the correct positions) and xor the resulting numbers together.
> I don't remember why this path was chosen, I think I considered it to be a good way to separate the code into stages since there are a lot of things going on at the same time.

Got your point. 
By "shift them accordingly to the argument's index (so that its bits end up at the correct positions) and xor the resulting numbers together", you mean the code below :
`            for (size_t j = 0; j < input_rows_count; ++j)
            {
                out_data[j] >>= 1;
                out_data[j] |= zShiftElement(
                        column_data.data + j * size_per_element,
                        size_per_element,
                        number_of_elements,
                        arg.type);
            }`
right ?

And it seems it still have problem. let's assume the argument is of type UInt8, the result type is UInt64, and we have 2 arguments.
Then, we only need to set the 8 bits in the result number. but, per the for loop here, we will set 64/2=32 bits.